packages/astro/src/vite-plugin-markdown/index.ts

```diff
@@ -167,19 +168,30 @@ export default function markdown({ config }: AstroPluginOptions): Plugin {
 				const prelude = `---
 import Slugger from 'github-slugger';
 ${layout ? `import Layout from '${layout}';` : ''}
-${components ? `import * from '${components}';` : ''}
+${isMDX && components ? `import * from '${components}';` : ''}
 ${hasInjectedScript ? `import '${PAGE_SSR_SCRIPT_ID}';` : ''}
-${setup}
+${isMDX ? setup : ''}
 
 const slugger = new Slugger();
 function $$slug(value) {
 	return slugger.slug(value);
 }
 
-const $$content = ${JSON.stringify(content)}
+const $$content = ${JSON.stringify(isMDX
+	? content
+	// Avoid stripping "setup" and "components"
+	// in plain MD mode
+	: { ...content, setup, components })}
 ---`;
 				const imports = `${layout ? `import Layout from '${layout}';` : ''}
-${setup}`.trim();
+${isMDX ? setup : ''}`.trim();
+
+				// Wrap with set:html fragment to skip
+				// JSX expressions and components in "plain" md mode
+				if (!isMDX) {
+					astroResult = `<Fragment set:html={${JSON.stringify(astroResult)}} />`
```

Comment:
Could also use `<Fragment is:raw>${astroResult}</Fragment>`? Not sure if the `set:html` is causing any problems here.

Reply:
That wouldn't process the HTML though, would it?

Reply:
It's effectively the same as what you have here, so no worries.